packages/react-dom/src/shared/ReactDOMFlushSync.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import {disableLegacyMode} from 'shared/ReactFeatureFlags';
+import {DiscreteEventPriority} from 'react-reconciler/src/ReactEventPriorities';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const ReactCurrentBatchConfig = ReactSharedInternals.ReactCurrentBatchConfig;
+
+import ReactDOMSharedInternals from 'shared/ReactDOMSharedInternals';
+const ReactDOMCurrentDispatcher =
+  ReactDOMSharedInternals.ReactDOMCurrentDispatcher;
+
+declare function flushSyncImpl<R>(fn: () => R): R;
+declare function flushSyncImpl(void): void;
+function flushSyncImpl<R>(fn: (() => R) | void): R | void {
+  const previousTransition = ReactCurrentBatchConfig.transition;
+  const previousEventPriority = ReactCurrentBatchConfig.eventPriority;
+
+  try {
+    ReactCurrentBatchConfig.transition = null;
```

Comment:
It's kind of unfortunate that we have to know about all the other options in every implementation. Because wrapping in one means disabling the outer one in either direction.

Although does that mean that startTransition should really reset eventPriority too?

Reply:
I think this happens automatically in that a transition supercedes the sync priority so we don't also have to change the priority. You can't really leave a transition except via another flushSync so I think it just works